vss-cats/src/main/scala/com/virtuslab/vss/cats/base/services/Passwords.scala

```diff
@@ -0,0 +1,106 @@
+      sql"select hash_type from hashed_passwords where password_hash = $passwordHash"
+        .query[String]
+        .to[List]
+        .transact(transactor)
+        .map(_.nonEmpty)
+    }
+
+    override def checkPwned(checkPwned: CheckPwned): F[CheckedPwned] = Trace[F].span("checkPwned") {
+      for {
+        pwnd <- doCheckPwned(checkPwned.passwordHash).map(CheckedPwned.apply(checkPwned.passwordHash, _))
+        _ <- Trace[F].put("passwordHash" -> checkPwned.passwordHash, "pwnd" -> pwnd.toString)
+        _ <- events.publishEvent(Event.CheckedPwned(checkPwned.passwordHash))
+      } yield pwnd
+    }
+
+    private def hashAlgorithm(hashType: String): F[(String => String)] = hashType match
+      case "sha256" => Monad[F].pure(DigestUtils.sha256Hex)
+      case _ =>
+        MonadThrow[F].raiseError(new RuntimeException("Unsupported hash type"))
+          <* Logger[F].info(s"Unsupported hash type: $hashType")
+
+    private def saveHash(hashedPassword: HashedPassword): F[Unit] = Trace[F].span("saveHash") {
+      sql"""|insert into hashed_passwords (password, hash_type, password_hash)
+            | values (${hashedPassword.password}, ${hashedPassword.hashType}, ${hashedPassword.hash}) on conflict do nothing""".stripMargin
+        .update
+        .run
+        .transact(transactor)
+        .void
+    }
+
+    private def getHash(hashType: String, password: String): F[Option[HashedPassword]] = Trace[F].span("getHash") {
+      sql"select hash_type, password, password_hash from hashed_passwords where hash_type = $hashType and password = $password"
+        .query[HashedPassword]
+        .option
+        .transact(transactor)
+    }
+
+    private def normalizeHashPassword(hashPassword: HashPassword): HashPassword =
+      hashPassword.focus(_.hashType).modify(_.toLowerCase)
+
+    override def hashPassword(_hashPassword: HashPassword): F[HashedPassword] = Trace[F].span("hashPassword") {
+      val hashPassword = normalizeHashPassword(_hashPassword)
+      for {
+        maybeHashedPassword <- getHash(hashPassword.hashType, hashPassword.password)
+        _ <- Trace[F].put("hashType" -> hashPassword.hashType, "password" -> hashPassword.password)
+        hashedPassword <- maybeHashedPassword match {
+          case Some(hashedPassword) =>
+            Monad[F].pure(hashedPassword)
+              <* Logger[F].info(s"Using cached hash for hashType = ${hashPassword.hashType} and password = ${hashPassword.password}}")
+          case None =>
+            for {
+              hashAlgorithm <- hashAlgorithm(hashPassword.hashType)
+              hash = hashAlgorithm(hashPassword.password)
+              hashedPassword = HashedPassword(hashPassword.hashType, hashPassword.password, hash)
+            } yield hashedPassword
+          }
+        _ <- Trace[F].put("hash" -> hashedPassword.hash)
+        _ <- events.publishEvent(Event.HashedPassword(hashPassword.password, hashPassword.hashType))
+        _ <- saveHash(hashedPassword)
```

Comment:
first save than publish event (in ideal word publishing event should be done reading the actually stored data in separate process)

Reply:
you mean the outbox pattern? yeah, might be a nice addition, or at least describe the guarantees of events publishing